File: medium/Two Sum II - Input Array Is Sorted/twoSum.js
/**
 * @param {number[]} numbers
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (numbers, target) {
  let start = 0;
  let end = numbers.length - 1;
  let sum = numbers[start] + numbers[end];

  while (sum != target) {
    if (sum > target) end -= 1;
    else start += 1;

    sum = numbers[start] + numbers[end];
  }

  return [start + 1, end + 1];
};

const testSum = function (nums, target, output) {
  const result = twoSum(nums, target);
  if (result == output) {
    console.log("Test Case Passed!");
  } else {
    console.log(
      `Test Case Failed: nums: ${nums} target: ${target} output: ${output} result: ${result}`
    );
  }
};

testSum([2, 7, 11, 15], 9, [1, 2]);
testSum([2, 3, 4], 6, [1, 3]);
testSum([-1, 0], -1, [1, 2]);
testSum([5, 25, 75], 100, [2, 3]);
testSum([3, 24, 50, 79, 88, 150, 345], 200, [3, 6]);

File: second_pass/2_two_pointers/2_two_sum/two_sum.py
from typing import List


class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        n = len(numbers)
        left = 0
        right = n - 1

        while (left <= right and left < n and right > 0):
            while left > 0 and numbers[left] == numbers[left - 1]:
                left += 1

            while right < n - 1 and numbers[right] == numbers[right + 1]:
                right -= 1

            calc = numbers[left] + numbers[right]

            if calc == target:
                return [left + 1, right + 1]

            if calc < target:
                left += 1

            if calc > target:
                right -= 1

    def testTwoSum(self, numbers, target, expected):
        actual = self.twoSum(numbers, target)
        if actual == expected:
            print("Test Case Passed!")
        else:
            print(f"Test Case Failed. numbers: {numbers} target: {target} actual: {actual} expected: {expected}")


Solution().testTwoSum([2, 7, 11, 15], 9, [1, 2])
Solution().testTwoSum([2, 3, 4], 6, [1, 3])
Solution().testTwoSum([-1, 0], -1, [1, 2])
